src/tools/compiletest/src/lib.rs

```diff
@@ -799,6 +799,23 @@ fn collect_tests_from_dir(
         return Ok(TestCollector::new());
     }
 
+    let mut components = dir.components().rev();
+    if let Some(Utf8Component::Normal(last)) = components.next()
+        && let Some(("assembly" | "codegen", backend)) = last.split_once('-')
+        && let Some(Utf8Component::Normal(parent)) = components.next()
+        && parent == "tests"
+        && let Ok(backend) = CodegenBackend::try_from(backend)
+        && backend != cx.config.codegen_backend
+    {
+        // We ignore asm tests which don't match the current codegen backend.
+        println!(
+            "Ignoring tests in `{dir}` because they don't match the configured codegen \
+             backend (`{}`)",
+            cx.config.codegen_backend.as_str(),
+        );
+        return Ok(TestCollector::new());
+    }
```

Comment:
Discussion: this feels *somewhat* surprising. Is this done because when you write `./x test assembly`, bootstrap will "splat" that to mean `./x test assembly-llvm assembly-gcc assembly-cranelift`?

Reply:
I think I might have another idea re. injecting `needs-codegen-backend` directives so tests show up as ignored and not have an empty test suite, but I'll play around with that separately. This is fine for now.